data/templates/src/pe/dll/template.c

```diff
@@ -20,82 +19,119 @@ void ExecutePayload(void);
 BOOL WINAPI
 DllMain (HANDLE hDll, DWORD dwReason, LPVOID lpReserved)
 {
-    switch (dwReason)
-    {
-        case DLL_PROCESS_ATTACH:
+	switch (dwReason)
+	{
+		case DLL_PROCESS_ATTACH:
 			ExecutePayload();
-            break;
+			break;
 
-        case DLL_PROCESS_DETACH:
-            // Code to run when the DLL is freed
-            break;
+		case DLL_PROCESS_DETACH:
+			// Code to run when the DLL is freed
+			break;
 
-        case DLL_THREAD_ATTACH:
-            // Code to run when a thread is created during the DLL's lifetime
-            break;
+		case DLL_THREAD_ATTACH:
+			// Code to run when a thread is created during the DLL's lifetime
+			break;
 
-        case DLL_THREAD_DETACH:
-            // Code to run when a thread ends normally.
-            break;
-    }
-    return TRUE;
+		case DLL_THREAD_DETACH:
+			// Code to run when a thread ends normally.
+			break;
+	}
+	return TRUE;
+}
+
+// Use a combination semaphore / event to check if the payload is already running and when it is, don't start a new
+// instance. This is to fix situations where the DLL is loaded multiple times into a host process and prevents the
+// payload from being executed multiple times. An event object is used to determine if the payload is currently running
+// in a child process. The event handle is created by this process (the parent) and configured to be inherited by the
+// child. While the child process is running, the event handle can be successfully opened. When the child process exits,
+// the event handle that was inherited from the parent will be automatically closed and subsequent calls to open it will
+// fail. This indicates that the payload is no longer running and a new instance can be created.
+BOOL Synchronize(void) {
+	BOOL bResult = TRUE;
+	BOOL bRelease = FALSE;
+	HANDLE hSemaphore = NULL;
+	HANDLE hEvent = NULL;
+	SECURITY_ATTRIBUTES SecurityAttributes;
+
+	// step 1: define security attributes that permit handle inheritance
+	SecurityAttributes.nLength = sizeof(SECURITY_ATTRIBUTES);
+	SecurityAttributes.lpSecurityDescriptor = NULL;
+	SecurityAttributes.bInheritHandle = TRUE;
+
+	do {
+		// step 2: create a semaphore to synchronize this routine
+		if ((hSemaphore = CreateSemaphoreA(&SecurityAttributes, 1, 1, szSyncNameS)) == NULL) {
+			// if the semaphore creation fails, break out using the default TRUE result, this shouldn't happen
+			break;
```

Comment:
Similar to the point above, if this fails, shouldn't `bResult` be `FALSE` as it failed? Right now this is going to cause this function to return `TRUE` the way its currently coded.

Reply:
After further discussion it was determined that this was intentional so as to allow code execution just in case something goes horribly wrong as we would rather gain a shell than block potential shells from occurring.